### DISCUSSION TITLE
Log request id

### DISCUSSION THREAD

**Discussion by reindert-vetter** [confetti-framework/confetti#121] [Ideas]
> Is there a plan to also refactor the CLI log once one fires up the application, for example, to be displaying the link also to the application [localhost:80](http://localhost:80) and also the log to be containing all requests made to your application?

Log request-id
Response with request-id header

_Originally posted by @KinyaElGrande in https://github.com/confetti-framework/confetti/discussions/105#discussioncomment-396039_

**Comment by bosix** [comment]
I was thinking about adding the request-id not only to the log messages but also to http responses. That way each part that has access to the request or the response can access the id. This also allows client-sided javascript to work 

Well, adding the request-id to request and response is really easy from my point of view. Here an example implementation with a middleware:

```golang
package middleware

import (
	"github.com/confetti-framework/contract/inter"
	"github.com/google/uuid"
)

type RequestID struct{}

const requestHeaderIDName = "x-request-id"

func (_ RequestID) Handle(request inter.Request, next inter.Next) inter.Response {
	var reqID string

        // allow other systems like a reverse proxy to set the request-id
	if reqID = request.Headers().Get(requestHeaderIDName); reqID == "" {
		reqID = uuid.New().String()
		request.Headers().Set(requestHeaderIDName, reqID)
	}

	resp := next(request)
	resp.GetHeaders().Set(requestHeaderIDName, reqID)

	return resp
}
```

My problem is that I have no clue how to access the request-id inside the logging facade (foundation -> logger_adapter.go). I think that this would be the correct spot to add the request-id globally to all log messages.

**Reply by reindert-vetter** [reply]
Nice example!

You're right. This is not possible now. I think in logger_adapter.go:92 we should iterate over custom decorators (callbacks). Multiple callbacks can then be defined in one field in Syslog.

An example:
``` patch
diff --git a/loggers/syslog.go b/loggers/syslog.go
index 7afd8f6..b89bccf 100644
--- a/loggers/syslog.go
+++ b/loggers/syslog.go
@@ -26,6 +26,7 @@ type Syslog struct {
        MaxFiles       int
        app            inter.AppReader
        HideStackTrace bool
+       Decorators     []func(d syslog.StructuredData) syslog.StructuredData
 }

 func (r Syslog) SetApp(app inter.AppReader) inter.Logger {
@@ -91,6 +92,10 @@ func (r Syslog) LogWith(severity log_level.Level, message string, rawContext int

        structuredData["level"] = syslog.SDElement{"severity": syslog.KeyBySeverity(severity)}

+       for _, decorator := range r.Decorators {
+               structuredData = decorator(structuredData)
+       }
+
        r.init().Log(
                severity,
                r.group,
```

Besides writing tests first, I do not think there needs to be much more code. What do you think of this?

**Reply by bosix** [reply]
I think we should provide 2 types of decorators: globally (will be applied to all log channels) and locally (will only be applied to one channel. Keep in mind that there is not only syslog as log-channel-type.

Further, I think we should not name it "decorators" since they do not change the format of the log message but are able to add more information. By the end of this week I'll have a deeper look at logging in confetti and maybe some other nameing comes in my mind. Maybe we should not make request ids optional at all or do you think that there is a use case where some users don't want to have them?

**Reply by reindert-vetter** [reply]
> we should provide 2 types of decorators

That is indeed true. Where do you use different channels for? At work, we only use slack, stdout and stderr. How do you use them other than how you can use `Group`?

> not only syslog as log-channel-type

You're right. I thought it would be difficult to make this global for all types, but that is actually not difficult at all. We could apply this in the following places:
loggers/logger_adapter.go:27
loggers/logger_adapter.go:31

(I want to rename `logger_adapter.go` to `logger_facade.go`)

> should not name it "decorators" 

Indeed, "decorators" is not the correct term. It also does not follow the decorator pattern.

> Maybe we should not make request ids optional at all

I would like it to be a standard middleware, but somehow optional (or highly customizable). At work we had to return a different format (for backwards compatibility). I think there will be many other middlewares in the future that we would like to include as standard, but optional.

**Reply by bosix** [reply]
> That is indeed true. Where do you use different channels for? At work, we only use slack, stdout and stderr. How do you use them other than how you can use Group?

At work, we do not send log messages directly to slack. All messages are sent to a log aggregator (e.g. graylog or logstash) to make some pre-processing. Since confetti uses `inter.Logger` as the expected logger interface a user could create a custom logger.

> We could apply this in the following places:
> loggers/logger_adapter.go:27
> loggers/logger_adapter.go:31

Sounds good. With that in place, custom loggers would also be affected and I think that's good.

> I think there will be many other middlewares in the future that we would like to include as standard, but optional.

That's a good point. There could be a list of all available optional middlewares inside the docs.

**Reply by reindert-vetter** [reply]
I don't know a nice name for it indeed.

I think I don't want to place those specific data providers on the channel (as below). Then we would have made the `inter.Logger` bigger with an extra method to retrieve those providers. What do you think of this?
``` patch
diff --git a/config/logging.go b/config/logging.go
index 7d78c96..969bfb5 100644
--- a/config/logging.go
+++ b/config/logging.go
@@ -11,6 +11,7 @@ import (
 var Logging = struct {
        Default  string
        Channels map[string]inter.Logger
+       DataProviders []func(map[string]interface{}) map[string]interface{}
 }{
        /*
           |--------------------------------------------------------------------------
@@ -24,6 +25,14 @@ var Logging = struct {
        */
        Default: "stack",

+       /**
+          | Global
+        */
+       DataProviders: []func(map[string]interface{}) map[string]interface{}{
+               loggers.appendLogLevel,
+               loggers.appendRequestId,
+       },
+
        /*
           |--------------------------------------------------------------------------
           | Log Channels
@@ -48,6 +57,10 @@ var Logging = struct {
                "single": loggers.Syslog{
                        Path:     Path.Storage + "/logs/default.log",
                        MinLevel: log_level.DEBUG,
+                       DataProviders: []func(map[string]interface{}) map[string]interface{}{
+                               loggers.appendExtraData,
+                               loggers.appendExtraData2,
+                       },
                },

                "daily": loggers.Syslog{
```

**Reply by bosix** [reply]
I see your point. What do you think about only allow global DataProviders/decorators/... ? That way the inter.Logger interface keeps thin and you could add information to the log channel. The only problem with that is when someone don't want to add a decorator to all channels.

**Reply by reindert-vetter** [reply]
Let's just make it global for now. Maybe we'll come up with a better idea later.

**Reply by bosix** [reply]
I've created a pr for adding the required middleware. Please have a look at the changes and let me know if you want me anything to add or change.
https://github.com/confetti-framework/foundation/pull/14

**Reply by reindert-vetter** [reply]
That looks good! I have 2 comments.
